backends/bmv2/synthesizeValidField.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "synthesizeValidField.h"
+
+#include <boost/range/combine.hpp>
+#include <boost/range/irange.hpp>
+
+#include "backends/bmv2/helpers.h"
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/methodInstance.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/p4/typeMap.h"
+#include "ir/ir.h"
+
+namespace BMV2 {
+
+/// Adds a valid bit to all header types.
+class AddValidField final : public Modifier {
+ public:
+    AddValidField() { setName("AddValidField"); }
+ private:
+    bool preorder(IR::Type_Header* header) override {
+        const cstring validField = V1ModelProperties::validField;
+        if (header->getField(validField) != nullptr) return true;
+        auto field = new IR::StructField(validField, IR::Type::Bits::get(1));
+        header->fields.push_back(field);
+        return true;
+    }
+};
+
+/// Replaces calls to isValid() with references to the header's valid bit.
```

Comment:
See the comments in the header. Currently we can't do that, because `setValid()` (when used in an action, at least) compiles down to the `add_header` primitive on BMV2. `add_header` does more than just set `$valid$` to 1 - for example, it's used in the implementation of header unions.